src/logic/function/basic.lean

```diff
@@ -357,6 +357,22 @@ rfl
 
 end bicomp
 
+section uncurry
+
+variables {α β γ δ : Type*}
+
+/-- Records a way to turn an element of `α` into a function from `β` to `γ`. -/
+class has_uncurry (α : Type*) (β : out_param Type*) (γ : out_param Type*) := (uncurry : α → (β → γ))
+
+notation `↿`:max x:max := has_uncurry.uncurry x
+
+instance has_uncurry_base : has_uncurry (α → β) α β := ⟨id⟩
+
+instance has_uncurry_induction [has_uncurry β γ δ] : has_uncurry (α → β) (α × γ) δ :=
+⟨λ f p, ↿(f p.1) p.2⟩
```

Comment:
Does mathlib have any guidelines for when to use `_` and when to use `.` in names? For instance, would it be more or less idiomatic to nest these under a `has_curry` namespace?

Reply:
Instances have their own rules for namespacing. They aren't usually written directly so it doesn't matter too much what they are called, but this one seems like it should be called `prod.has_uncurry` or similar.

@PatrickMassot , this instance violates uniqueness of instances: since the last two parameters are `out_param`, they should be determined by the first parameter, but both instances clearly apply to an implication.

Reply:
@digama0: apologies for derailing with beginner questions, but... Where can I learn about what `out_param` does?

Reply:
There isn't a good reference AFAIK, but they are the same as "output functional dependencies" in haskell. See https://en.wikibooks.org/wiki/Haskell/Advanced_type_classes#Functional_dependencies , or ask on zulip and I can say more.

Reply:
I don't understand your implication story. `Prop` isn't allowed anywhere here. And those instances (and more) are used without any issue in the sphere eversion project.

Reply:
I think @digama0 meant `α → β` by "implication". I too wonder why it works: `out_param`s should be determined by the first argument which is `α → β`. You declare two instances with the same first argument. I guess Lean tries the "induction" instance first. I think that @gebner knows better than us how it works.

Reply:
@digama0 Incoherence is very much intentional here, because we need to distinguish between `α × β` and `γ`.

Reply:
Perhaps there should be some priorities set, to make it explicit then?

Reply:
I'm also a little confused by the story here and it would be nice to see it documented, perhaps with explicit priorities like @digama0 suggests.